core/src/main/cpp/gdn/projection_mesh/projection_mesh.h

```diff
@@ -0,0 +1,213 @@
+#ifndef PROJECTION_MESH_H
+#define PROJECTION_MESH_H
+
+#include <core/Ref.hpp>
+#include <core/Vector2.hpp>
+#include <gen/ExternalTexture.hpp>
+#include <gen/MeshInstance.hpp>
+#include <gen/Resource.hpp>
+#include <gen/Shader.hpp>
+#include <gen/ShaderMaterial.hpp>
+#include <gen/Shape.hpp>
+
+#include "projection_mesh_utils.h"
+#include "utils.h"
+
+namespace gast {
+
+namespace {
+using namespace godot;
+const char *kGastTextureParamName = "gast_texture";
+const char *kGastEnableBillBoardParamName = "enable_billboard";
+const char *kGastGradientHeightRatioParamName = "gradient_height_ratio";
+const char *kGastNodeAlphaParamName = "node_alpha";
+const int kDefaultSurfaceIndex = 0;
+const Vector2 kInvalidCoordinate = Vector2(-1, -1);
+}
+
+class ProjectionMesh : public Resource {
+GODOT_CLASS(ProjectionMesh, Resource)
+
+public:
+    ProjectionMesh();
+    ~ProjectionMesh();
+
+    void _init();
+
+    void _process(const real_t delta);
+
+    static void _register_methods();
+
+    // Mirrors src/main/java/org/godotengine/plugin/gast/GastNode#ProjectionMeshType
+    enum ProjectionMeshType {
+        RECTANGULAR = 0,
+        EQUIRECTANGULAR = 1,
+        MESH = 2,
+    };
+
+    inline ProjectionMeshType get_projection_mesh_type() const {
+        return projection_mesh_type;
+    }
+
+    inline bool is_rectangular_projection_mesh() const {
+        return get_projection_mesh_type() == ProjectionMeshType::RECTANGULAR;
+    }
+
+    inline bool is_equirectangular_projection_mesh() const {
+        return get_projection_mesh_type() == ProjectionMeshType::EQUIRECTANGULAR;
+    }
+
+    inline MeshInstance* get_mesh_instance() {
+        return this->mesh_instance;
+    }
+
+    inline Ref<Shape> get_collision_shape() {
+        return this->collision_shape_ref;
+    }
+
+    inline void set_external_texture(Ref<ExternalTexture> external_texture) {
+        if (!shader_material_ref.is_valid()) {
+            ALOGE("Cannot set ExternalTexture, invalid ShaderMaterial.");
+            return;
+        }
+        shader_material_ref->set_shader_param(kGastTextureParamName, external_texture);
+    }
+
+    inline ExternalTexture* get_external_texture() {
+        if (!shader_material_ref.is_valid()) {
+            return nullptr;
+        }
+
+        ExternalTexture *external_texture = nullptr;
+        Ref<Texture> texture = shader_material_ref->get_shader_param(kGastTextureParamName);
+        if (texture.is_valid()) {
+            external_texture = Object::cast_to<ExternalTexture>(*texture);
+        }
+        return external_texture;
+    }
+
+    inline void set_render_on_top(bool enable) {
+        if (this->render_on_top == enable) {
+            return;
+        }
+        this->render_on_top = enable;
+        if (shader_material_ref.is_valid() && shader_material_ref->get_shader().is_valid()) {
+            String shader_code = shader_material_ref->get_shader()->get_code();
+            shader_code += kDisableDepthTestRenderMode;
```

Comment:
This should check for the value of `render_on_top`.
Also the line above seems to imply that the current shader code is retrieved, then updated. What if `render_on_top` was `true` and then being set to `false`. In that case you'd want the original shader code without the addition.

Reply:
Switched to using generate_shader_code to regenerate the code here, which should do the right thing. Tested it by setting an equirect projection to render on top, then turning it off. Both turning on and off worked.